pandas/io/formats/string.py

```diff
@@ -0,0 +1,199 @@
+        return text
+
+    def _get_strcols(self) -> List[List[str]]:
+        strcols = self.fmt.get_strcols()
+        if self.fmt.is_truncated:
+            strcols = self._insert_dot_separators(strcols)
+        return strcols
+
+    def _get_string_representation(self) -> str:
+        if self.fmt.frame.empty:
+            return self._empty_info_line
+
+        strcols = self._get_strcols()
+
+        if self.line_width is None:
+            # no need to wrap around just print the whole frame
+            return self.adj.adjoin(1, *strcols)
+
+        if self._need_to_wrap_around:
+            return self._join_multiline(strcols)
+
+        return self._fit_strcols_to_terminal_width(strcols)
+
+    @property
+    def _empty_info_line(self) -> str:
+        return (
+            f"Empty {type(self.frame).__name__}\n"
+            f"Columns: {pprint_thing(self.frame.columns)}\n"
+            f"Index: {pprint_thing(self.frame.index)}"
+        )
+
+    @property
+    def _need_to_wrap_around(self) -> bool:
+        return bool(self.fmt.max_cols is None or self.fmt.max_cols > 0)
+
+    def _insert_dot_separators(self, strcols: List[List[str]]) -> List[List[str]]:
+        str_index = self.fmt._get_formatted_index(self.fmt.tr_frame)
+        index_length = len(str_index)
+
+        if self.fmt.is_truncated_horizontally:
+            strcols = self._insert_dot_separator_horizontal(strcols, index_length)
+
+        if self.fmt.is_truncated_vertically:
+            strcols = self._insert_dot_separator_vertical(strcols, index_length)
+
+        return strcols
+
+    def _insert_dot_separator_horizontal(
+        self, strcols: List[List[str]], index_length: int
+    ) -> List[List[str]]:
+        strcols.insert(self.fmt.tr_col_num + 1, [" ..."] * index_length)
+        return strcols
+
+    def _insert_dot_separator_vertical(
+        self, strcols: List[List[str]], index_length: int
+    ) -> List[List[str]]:
+        n_header_rows = index_length - len(self.fmt.tr_frame)
+        row_num = self.fmt.tr_row_num
+        for ix, col in enumerate(strcols):
+            cwidth = self.adj.len(col[row_num])
+
+            if self.fmt.is_truncated_horizontally:
+                is_dot_col = ix == self.fmt.tr_col_num + 1
+            else:
+                is_dot_col = False
+
+            if cwidth > 3 or is_dot_col:
+                dots = "..."
+            else:
+                dots = ".."
+
+            if ix == 0:
+                dot_mode = "left"
+            elif is_dot_col:
+                cwidth = 4
+                dot_mode = "right"
+            else:
+                dot_mode = "right"
+
+            dot_str = self.adj.justify([dots], cwidth, mode=dot_mode)[0]
+            col.insert(row_num + n_header_rows, dot_str)
+        return strcols
+
+    def _join_multiline(self, strcols: List[List[str]]) -> str:
+        lwidth = self.line_width
+        adjoin_width = 1
+        if self.fmt.index:
+            idx = strcols.pop(0)
```

Comment:
AFAICT the original _join_multiline didn't mutate the input.

often better to type inputs as Iterable instead of List for non-mutating functions

Reply:
Oh, you are right! It did not mutate strcols, but now it does. I will fix it.

What I don't like about Iterable type is the following:
- String is Iterable, so it can be confused with List/Tuple.
- Cannot find len(iterable) - mypy throws
```Argument 1 to "len" has incompatible type "Iterable[str]"; expected "Sized"  [arg-type]```

So, now I changed the type of input parameter in ``_join_multiline`` to ``Iterable[List[str]]``.

Reply:
> * String is Iterable, so it can be confused with List/Tuple.

yeah that can be a problem, but we normally add type parameters, so only Iterable[str] becomes an issue

> * Cannot find len(iterable) - mypy throws

can always use Sequence if needed as Sequence is also not mutable.

There are some guidelines for stubs at https://github.com/python/typeshed/blob/master/CONTRIBUTING.md#stub-file-coding-style

> avoid invariant collection types (List, Dict) in argument positions, in favor of covariant types like Mapping or Sequence

at some point, the guidelines there relevant to the codebase and not just stubs should probably be added to our code style docs #33851